### PR TITLE
Add multiple file support

### PR DESCRIPTION
**Pull request by dgzlopes**
Signed-off-by: Daniel González Lopes <danielgonzalezlopes@gmail.com>

Small fix for #180. Multiple file support might be interesting for the new benchmark script.

### CODE REVIEW

File: detect_secrets/core/baseline.py [hunk 1]
```diff
@@ -12,7 +12,7 @@ def initialize(
     plugins,
     exclude_files_regex=None,
     exclude_lines_regex=None,
-    path='.',
```
Comment (domanchi) [inline detect_secrets/core/baseline.py]: Generally speaking, it's a bad idea to have a list as a default parameter, due to the following reason:

```
def foobar(L=['a']):
    L.append('b')
    print(L)

foobar()    # ['a', 'b']
foobar()    # ['a', 'b', 'b']
```

While we may not be appending to `path` here, we're introducing a "gotcha" that may come to bite us later on. A better way of doing this is:

```
def foobar(L=None):
    if not L:
        L = []
```

And since we're doing this, we can also do:

```
def foobar(L='.'):
    if not isinstance(L, list):
        L = [L]
```

which provides a nicer API for us to use (rather than always calling the function with a list of length 1).

File: detect_secrets/core/baseline.py [hunk 2]
```diff
@@ -34,13 +34,15 @@ def initialize(
         exclude_lines=exclude_lines_regex,
     )
 
-    if os.path.isfile(path):
+    if len(path) > 1 or os.path.isfile(path[0]):
         # This option allows for much easier adhoc usage.
-        files_to_scan = [path]
+        files_to_scan = list()
+        for elem in path:
+            files_to_scan.append(elem)
```
Comment (domanchi) [inline detect_secrets/core/baseline.py]: Since `path` is already a list, isn't this just:

```
files_to_scan = path
```

?

File: detect_secrets/core/baseline.py [hunk 3]
```diff
@@ -34,13 +34,15 @@ def initialize(
         exclude_lines=exclude_lines_regex,
     )
 
-    if os.path.isfile(path):
+    if len(path) > 1 or os.path.isfile(path[0]):
```
Comment (domanchi) [inline detect_secrets/core/baseline.py]: I'm doubtful this carries over the same logic as the previous version.

After all, this should allow invocations like:

```
$ detect-secrets scan non-existent-file.A non-existent-file.B
```

right?

File: detect_secrets/core/baseline.py (hunk 2 above)
Comment (dgzlopes) [reply detect_secrets/core/baseline.py]: Completely missed this! Sorry, was part of some behavior testing but It's removed now.

File: detect_secrets/core/baseline.py (hunk 1 above)
Comment (dgzlopes) [reply detect_secrets/core/baseline.py]: Thanks for pointing this :) Tried to solve It on the last update, looks better now?

File: detect_secrets/core/baseline.py (hunk 3 above)
Comment (dgzlopes) [reply detect_secrets/core/baseline.py]: Took a closer look at this and to be fair I think that I misinterpreted what the previous version was doing. So updated it to:
- Running `detect-secrets scan non-existent-file.A non-existent-file.B`  skips the invocation.
- Running `detect-secrets scan` or `detect-secrets scan detect-secrets` works as expected.
- Running `detect-secrets scan non-existent-file.A existent-file.B`  skips the invocation too! But I'm not sure if this is great or not. If one argument isn't a file scanning the other ones should work?

Do you think this is the proper way to solve this?

**Comment by dgzlopes** [comment]
@domanchi Added `multiple valid files` and  `multiple invalid files` tests on the last update! Should I add test for some more edge case?

And thank you for taking some time to review the PR!

File: tests/main_test.py [hunk 4]
```diff
@@ -41,7 +41,7 @@ def test_scan_with_rootdir(self, mock_baseline_initialize):
             plugins=Any(tuple),
             exclude_files_regex=None,
             exclude_lines_regex=None,
-            path='test_data',
```
Comment (dgzlopes) [inline tests/main_test.py]: I don't understand why tests won't work if I remove this. Running `detect-secrets scan test_data` seems to work properly.

File: detect_secrets/core/baseline.py (hunk 3 above)
Comment (domanchi) [reply detect_secrets/core/baseline.py]: That's a good question you raise. When it comes to command line tools usability, I defer back to tried-and-true unix tools:

```
detect-secrets $ cat asdf
cat: asdf: No such file or directory
detect-secrets $
detect-secrets $ cat asdf requirements-dev.txt
cat: asdf: No such file or directory
coverage
flake8==3.5.0
mock
monotonic
pre-commit==1.11.2
pytest
pyyaml
tox>=3.8
tox-pip-extensions
unidiff
```

Perhaps we should do something similar? You can print to stderr, or you can just use:

```
log.error(...)
```

File: detect_secrets/core/baseline.py [hunk 5]
```diff
@@ -34,13 +34,19 @@ def initialize(
...
-    if os.path.isfile(path):
+    if not isinstance(path, list):
+        path = [path]
+
+    if check_all_are_files(path):
         # This option allows for much easier adhoc usage.
-        files_to_scan = [path]
-    elif scan_all_files:
-        files_to_scan = _get_files_recursively(path)
+        files_to_scan = path
```
Comment (domanchi) [inline detect_secrets/core/baseline.py]: If we decide to pursue the option to display errors for non existing files, we should probably change this to a filter.

File: detect_secrets/core/baseline.py [hunk 6]
```diff
@@ -34,13 +34,19 @@ def initialize(
         exclude_lines=exclude_lines_regex,
     )
 
-    if os.path.isfile(path):
```
Comment (domanchi) [inline detect_secrets/core/baseline.py]: Another reason for the original logic was that it allowed us to do:

```
detect-secrets scan test_data/files/file_with_secrets.py
```

as well as

```
detect-secrets scan test_data/
```

using `_get_git_tracked_files` to get files, if it is a directory.

How does this change handle that?

**Comment by dgzlopes** [comment]
Learned a lot from the review! Both about Python and form the problem. So I tried to fix the missing logic and all the small bits. Now for each element on the path, I check if It's a directory or a file: 

-If It's a directory, in case we have the `scan_all_files` flag active It runs` _get_files_recursively(element)`... If not It runs `_get_git_tracked_files(element)`. In both cases, It extends the returned list to the `files_to_scan` list.
- If it's a file It's appended to the` files_to_scan` list.
- If the element doesn't exist the error is logged!

File: detect_secrets/core/baseline.py (hunk 6 above)
Comment (dgzlopes) [reply detect_secrets/core/baseline.py]: Should be working now :)

File: detect_secrets/core/baseline.py (hunk 5 above)
Comment (dgzlopes) [reply detect_secrets/core/baseline.py]: On the new update didn't use the filter as I changed the whole logic! With the new changes, It's still needed? As I'm not checking for errors on the whole list...

File: detect_secrets/core/baseline.py [hunk 7]
```diff
@@ -60,6 +66,13 @@ def initialize(
     return output
 
 
+def check_all_are_files(path):
+    for elem in path:
+        if os.path.isfile(elem) is False:
```
[the remark this answers is not in the document]
Comment (dgzlopes) [reply detect_secrets/core/baseline.py]: Thanks for the point!

File: tests/core/baseline_test.py [hunk 8]
```diff
@@ -58,6 +58,23 @@ def test_basic_usage(self, path):
...
+    def test_with_multiple_files(self):
+        results = self.get_results(path=[
+            'test_data/files/file_with_secrets.py',
+            'test_data/files/tmp/file_with_secrets.py',
+        ])
+
+        assert len(results['test_data/files/file_with_secrets.py']) == 1
+        assert len(results['test_data/files/tmp/file_with_secrets.py']) == 2
+        assert 'test_data/files/file_with_secrets.py' in results
+        assert 'test_data/files/tmp/file_with_secrets.py' in results
```
[the remark this answers is not in the document]
Comment (dgzlopes) [reply tests/core/baseline_test.py]: Added test case with a mix of folders and files and some negative asserts that check if some files are or not in the results!

File: detect_secrets/core/baseline.py [hunk 9]
```diff
@@ -23,7 +23,7 @@ def initialize(
 
     :type exclude_files_regex: str|None
     :type exclude_lines_regex: str|None
-    :type path: str
+    :type path: list
```
[the remark this answers is not in the document]
Comment (dgzlopes) [reply detect_secrets/core/baseline.py]: Fixed!

File: detect_secrets/core/baseline.py [hunk 10]
```diff
@@ -34,13 +34,19 @@ def initialize(
...
+    if check_all_are_files(path):
         # This option allows for much easier adhoc usage.
-        files_to_scan = [path]
-    elif scan_all_files:
-        files_to_scan = _get_files_recursively(path)
+        files_to_scan = path
     else:
-        files_to_scan = _get_git_tracked_files(path)
+        if len(path) > 1:
+            return output
```
[the remark this answers is not in the document]
Comment (dgzlopes) [reply detect_secrets/core/baseline.py]: It should be working now!

File: tests/main_test.py (hunk 4 above)
Comment (dgzlopes) [reply tests/main_test.py]: Ok! Thanks for explaining :)

File: detect_secrets/core/baseline.py (hunk 3 above)
Comment (dgzlopes) [reply detect_secrets/core/baseline.py]: Added error logging!

File: tests/core/baseline_test.py [hunk 11]
```diff
@@ -58,6 +58,30 @@ def test_basic_usage(self, path):
...
+    def test_with_multiple_non_existent_files(self):
+        results = self.get_results(path=['non-existent-file.A', 'non-existent-file.B'])
+
+        # No expected results, because files don't exist
+        assert not results
+
+    def test_with_folders_and_files(self):
+        results = self.get_results(path=['test_data/', 'non-existent-file.B'])
+
+        assert len(results.keys()) == 10
```
Comment (domanchi) [inline tests/core/baseline_test.py]: Is there a better way you can do this?

My only concern for this magic number is:

1. Another developer creates a new test file under `test_data/`
2. They write a new test, that leverages that new file. This works.
3. For some bizarre reason, this test fails now.

The CI will catch failing tests, but it's just less convenient if there's a better option.

File: detect_secrets/core/usage.py [hunk 12]
```diff
@@ -120,7 +120,7 @@ def add_arguments(self):
     def _add_initialize_baseline_argument(self):
         self.parser.add_argument(
             'path',
-            nargs='?',
+            nargs='*',
```
Comment (KevinHock) [inline detect_secrets/core/usage.py]: I'm not sure, but would it be possible to use `'+'`? So you wouldn't have to do
```python
:type path: str|list
```
and the `isinstance` check

**Comment by MVrachev** [comment]
I believe it will be useful to add information about this feature in the README.
Maybe just a small example calling `detect-secret` on multiple files.

File: tests/core/baseline_test.py (hunk 11 above)
Comment (dgzlopes) [reply tests/core/baseline_test.py]: Added a couple of known asserts (to check that directory scan works as expected) and removed the hardcoded test!

File: detect_secrets/core/usage.py (hunk 12 above)
Comment (dgzlopes) [reply detect_secrets/core/usage.py]: Well, as '+' generates an error message if there isn't at least one command-line argument present... I thought that as we are using default's, '*' would be a better fit. Running the test suite with '+' breaks with `detect-secrets scan: error: too few arguments` (Well, as intended!).

Maybe I'm missing something... But If you think it's an improvement I can take a closer look at it :)

**Comment by dgzlopes** [comment]
@MVrachev I thought that multiple file support was an obvious feature for this type of tool! But I added a little Unix-like doc to the readme explaining `scan` mode inputs as this might not be that obvious :)

File: detect_secrets/core/usage.py (hunk 12 above)
Comment (KevinHock) [reply detect_secrets/core/usage.py]: You’re completely right :) 

From reading https://stackoverflow.com/questions/23490152/list-of-arguments-with-argparse#23490179 it seems it should be always be a list with *, so maybe that’s true?

Comment (dgzlopes) [reply detect_secrets/core/usage.py]: Yes! Looks like all is passed as a list, even the default argument ` '.' `

So the isInstance() might be redundant as when we pass (on the initialize()) `path='.' `  it's never used... Running scan without path(s) will trigger argparse's default. 

This was closely related to https://github.com/Yelp/detect-secrets/pull/188#discussion_r288614772 (@domanchi) so maybe using the list as a default isn't that great... But I think '+' passes all as a list too! What's the best way to handle this?

Comment (KevinHock) [reply detect_secrets/core/usage.py]: If I understand, the 

```python
    if not isinstance(path, list):
        path = [path]
```
is always dead-code, so we can remove that. 

From reading the code, it seems a path is always passed in to `initialize`, so there doesn't seem to be a need for a default argument I think, right?

If so we can change
```python
def initialize(
    plugins,
    exclude_files_regex=None,
    exclude_lines_regex=None,
    path='.',
    scan_all_files=False,
):
```
to
```
def initialize(
    plugins,
    path,
    exclude_files_regex=None,
    exclude_lines_regex=None,
    scan_all_files=False,
):
```

Comment (dgzlopes) [reply detect_secrets/core/usage.py]: As I thought :) Thanks for the writeup. Updated the PR and fixed the failing tests.

Comment (KevinHock) [reply detect_secrets/core/usage.py]: Looks awesome 👏
